### PR TITLE
Add register function for migration cache

### PR DESCRIPTION
**Pull request by maggie-lou**
---

**Version bump**: TODO

### CODE REVIEW

**Comment by maggie-lou** [comment]
Current dependencies on/for this PR:
* refactor_cache_options
  * **PR #2485** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/2485" target="_blank">Graphite</a> 
    * **PR #2486** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/2486" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/2486?utm_source=stack-comment).

File: server/backends/migration_cache/migration_cache.go [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+func validateCacheConfig(config CacheConfig) error {
+	numConfigs := 0
+	v := reflect.ValueOf(config)
+	for i := 0; i < v.NumField(); i++ {
+		if !v.Field(i).IsNil() {
+			numConfigs++
+		}
+	}
```
Comment (tylerwilliams) [inline server/backends/migration_cache/migration_cache.go:64]: Kinda gnarly to use reflection here! Can you just write this code out?

File: server/backends/migration_cache/migration_cache.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	c.Start()
+	env.GetHealthChecker().RegisterShutdownFunction(func(ctx context.Context) error {
+		return c.Stop()
+	})
+	return c, nil
+}
+
+func (mc MigrationCache) WithIsolation(ctx context.Context, cacheType interfaces.CacheType, remoteInstanceName string) (interfaces.Cache, error) {
+	//TODO implement me
+	panic("implement me")
```
Comment (tylerwilliams) [inline server/backends/migration_cache/migration_cache.go:133]: nit: I know this is placeholder code but typically we don't use panics in server code ever.

You could instead

```
return nil, status.UnimplementedError("not yet implemented")
```

**Review by tylerwilliams — CHANGES_REQUESTED** [review]

File: server/backends/migration_cache/migration_cache.go (hunk 2 above)
Comment (maggie-lou) [reply server/backends/migration_cache/migration_cache.go]: Done

File: server/backends/migration_cache/migration_cache.go (hunk 1 above)
Comment (maggie-lou) [reply server/backends/migration_cache/migration_cache.go]: Yeah thought that would make it more generalizable, but agree that it's probably just over-engineering

File: server/backends/migration_cache/config.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+type MigrationConfig struct {
+	Src  *CacheConfig `yaml:"src"`
+	Dest *CacheConfig `yaml:"dest"`
+}
+
+type CacheConfig struct {
+	DiskConfig   *DiskCacheConfig   `yaml:"disk"`
+	PebbleConfig *PebbleCacheConfig `yaml:"pebble"`
+}
```
Comment (tempoz) [inline server/backends/migration_cache/config.go:17]: Idea: make this an interface that defines "CacheFromConfig() interfaces.Cache`, and then you don't have to validate the CacheConfig object.

**Review by tylerwilliams — APPROVED** [review]

File: server/backends/migration_cache/config.go (hunk 3 above)
Comment (maggie-lou) [reply server/backends/migration_cache/config.go:17]: Can you say more about how to do that in a way where we can still enforce the structure of what we're expecting of a DiskCacheConfig, for example?

Comment (maggie-lou) [reply server/backends/migration_cache/config.go:17]: Going to merge this for now bc that will make some other development easier, but can come back and refactor this

Comment (tempoz) [reply server/backends/migration_cache/config.go:17]: The idea is that you make `DiskCacheConfig` implement `CacheConfig`, and then whatever gets passed in as a `CacheConfig` is guaranteed to just be one concrete type of `CacheConfig`, so it doesn't need to be validated. Also if we want to add a new type later, it simply has to implement `CacheConfig` and there are no changes that need to be made elsewhere, either in validation or in `CacheConfig`'s (now non-existent) fields.

Comment (maggie-lou) [reply server/backends/migration_cache/config.go:17]: I like the concept but does that work with the JSON struct reflection you helped me with last week? We'd need the yaml parser to figure out which concrete type the interface needs to be cast to right?

Comment (maggie-lou) [reply server/backends/migration_cache/config.go:17]: I tried this, which is what I think you're suggesting: https://github.com/buildbuddy-io/buildbuddy/pull/2507/files

Getting errors like: 	panic: reflect.Set: value of type map[string]interface {} is not assignable to type migration_cache.CacheConfig

**Pull request merged**